optapy-core/src/main/python/test/__init__.py

```diff
@@ -0,0 +1,609 @@
+from typing import Callable, Generic, List, Type, TypeVar, TYPE_CHECKING, overload, Union
+from jpype import JProxy
+
+from ..jpype_type_conversions import PythonBiFunction
+from ..optaplanner_java_interop import get_class
+from ..constraint_stream import PythonConstraintFactory, BytecodeTranslation
+
+if TYPE_CHECKING:
+    # These imports require a JVM to be running, so only import if type checking
+    from org.optaplanner.core.api.score.stream import Constraint, ConstraintFactory
+    from org.optaplanner.core.config.solver import SolverConfig
+    from org.optaplanner.core.api.score import Score
+
+
+Solution_ = TypeVar('Solution_')
+
+
+class PythonConstraintVerifier(Generic[Solution_]):
+    """
+    Entry point for the ConstraintVerifier API, which is used to test constraints defined by
+    a @constraint_provider function.
+    """
+    def __init__(self, delegate):
+        self.delegate = delegate
+        self.bytecode_translation = BytecodeTranslation.IF_POSSIBLE
+
+    def with_bytecode_translation(self, bytecode_translation: BytecodeTranslation) ->\
+            'PythonConstraintVerifier[Solution_]':
+        """
+        All subsequent calls to verify_that(constraint_function) will translate bytecode according to the rules
+        of the given BytecodeTranslation
+
+        :param bytecode_translation: A BytecodeTranslation member.
+        :return: self, for chaining
+        """
+        self.bytecode_translation = bytecode_translation
+        return self
+
+    @overload
```

Comment:
I'm a bit confused here. I don't see `PythonConstraintVerifier` extend or implement anything, yet there is apparently an overloaded method. What does it overload, from where?

Also, shouldn't `PythonConstraintVerifier` somehow implement OP's `ConstraintVerifier`? How do you ensure that the APIs are in sync?

Reply:
In Python, `typing.overload` (here imported as `overload`) is used to mark a function definition as a particular function signature for a more broad function. For instance:
```python
@typing.overload
def my_fun(a: int):
    ...

@typing.overload
def my_fun(msg: str):
    ...

@typing.overload
def my_fun(a: int, msg: str):
    ...

def my_fun(a: int | str, msg: str):
    # actual implementation goes here
    ...
```
declares `my_fun` to have three valid signatures: `int`, `str` and `int, str`. Do note Java overloads do not exist in the Python world; a function can only have one signature (which is the most broad one). IDE and type checkers use `typing.overload` to determine return types and parameter types for broad functions.

Reply:
In regards to `PythonConstraintVerifier` somehow reference OP's `ConstraintVerifier`, that can be done with `@JImplements` (which will raise an error if the API's are not in sync). However, there might be an issue if methods are overloaded (for instance, `verifyThat`) (need to verify).